hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/monitoring/MonitoringCache.java

```diff
@@ -0,0 +1,180 @@
+    }
+
+    private void checkSubscriptionsPartitions() {
+        if (usingCache == 1) {
+            subscriptionsWithUnassignedPartitions2.clear();
+            usingCache = 2;
+        } else {
+            subscriptionsWithUnassignedPartitions1.clear();
+            usingCache = 1;
+        }
+        monitorSubscriptionsPartitions();
+    }
+
+    private void monitorSubscriptionsPartitions() {
+        List<TopicSubscription> topicSubscriptions = getAllActiveSubscriptions();
+        List<List<TopicSubscription>> splitSubscriptions = splitSubscriptions(topicSubscriptions);
+
+        ExecutorService executorService = Executors.newFixedThreadPool(monitoringProperties.getNumberOfThreads());
+
+        for (int thread = 0; thread < monitoringProperties.getNumberOfThreads(); thread++) {
+            int part = thread;
+            executorService.submit(() -> {
+                try {
+                    List<MonitoringService> monitoringServices = createMonitoringService();
+                    List<TopicSubscription> subscriptionsWithUnassignedPartitions = new ArrayList<>();
+                    logger.info("Monitoring {} started for {} subscriptions", part, splitSubscriptions.get(part).size());
+
+                    splitSubscriptions.get(part).forEach(topicSubscription ->
+                            monitoringServices.forEach(monitoringService -> {
+                                        if (!monitoringService.checkIfAllPartitionsAreAssigned(topicSubscription.getTopic(),
+                                                topicSubscription.getSubscription())) {
+                                            subscriptionsWithUnassignedPartitions.add(topicSubscription);
+                                        }
+                                    }
+                            ));
+                    addSubscriptionsToCache(subscriptionsWithUnassignedPartitions);
+                } catch (Exception e) {
+                    logger.error("Error in monitoring: ", e);
+                } finally {
+                    logger.info("Monitoring {} ended", part);
+                    executorService.shutdown();
+                }
+            });
+        }
+    }
+
+    private void addSubscriptionsToCache(List<TopicSubscription> subscriptionsWithUnassignedPartitions) {
+        if (usingCache == 1) {
+            subscriptionsWithUnassignedPartitions1.add(subscriptionsWithUnassignedPartitions);
+        } else {
+            subscriptionsWithUnassignedPartitions2.add(subscriptionsWithUnassignedPartitions);
+        }
+    }
+
+    private List<List<TopicSubscription>> splitSubscriptions(List<TopicSubscription> topicSubscriptions) {
+        List<List<TopicSubscription>> splitSubscriptions = new ArrayList<>(Collections.emptyList());
+        splitSubscriptions.add(topicSubscriptions.subList(0, topicSubscriptions.size() / monitoringProperties.getNumberOfThreads()));
+        for (int i = 1; i < monitoringProperties.getNumberOfThreads(); i++) {
```

Comment:
What do You think about using something common to split a list, for example [this](https://stackoverflow.com/questions/12026885/is-there-a-common-java-utility-to-break-a-list-into-batches) instead of developing our own algorithm for this?

Reply:
Cool, thanks :D